objects/src/accounts/account_id.rs

```diff
@@ -134,336 +108,511 @@ impl FromStr for AccountStorageMode {
 
     // CONSTRUCTORS
     // --------------------------------------------------------------------------------------------
 
-    /// Returns a new account ID derived from the specified seed, code commitment and storage
-    /// commitment.
+    /// Creates an [`AccountId`] by hashing the given `seed`, `code_commitment`,
+    /// `storage_commitment` and [`AccountIdAnchor::block_hash`] from the `anchor` and using the
+    /// resulting first and second element of the hash as the first and second felt of the ID.
+    /// The [`AccountIdAnchor::epoch`] from the `anchor` overwrites part of the second felt.
     ///
-    /// The account ID is computed by hashing the seed, code commitment and storage commitment and
-    /// using 1 element of the resulting digest to form the ID. Specifically we take element 0.
-    /// We also require that the last element of the seed digest has at least `23` trailing
-    /// zeros if it is a regular account, or `31` trailing zeros if it is a faucet account.
+    /// Note that the `anchor` must correspond to a valid block in the chain for the ID to be deemed
+    /// valid during creation.
     ///
-    /// The seed digest is computed using a sequential hash over
-    /// hash(SEED, CODE_COMMITMENT, STORAGE_COMMITMENT, ZERO).  This takes two permutations.
+    /// See the documentation of the [`AccountId`] for more details on the generation.
     ///
     /// # Errors
-    /// Returns an error if the resulting account ID does not comply with account ID rules:
-    /// - the metadata embedded in the ID (i.e., the first 4 bits) is valid.
-    /// - the ID has at least `5` ones.
-    /// - the last element of the seed digest has at least `23` trailing zeros for regular accounts.
-    /// - the last element of the seed digest has at least `31` trailing zeros for faucet accounts.
+    ///
+    /// Returns an error if any of the ID constraints are not met. See the [type
+    /// documentation](AccountId) for details.
     pub fn new(
         seed: Word,
+        anchor: AccountIdAnchor,
         code_commitment: Digest,
         storage_commitment: Digest,
     ) -> Result<Self, AccountError> {
-        let seed_digest = compute_digest(seed, code_commitment, storage_commitment);
+        let seed_digest =
+            compute_digest(seed, code_commitment, storage_commitment, anchor.block_hash());
 
-        Self::validate_seed_digest(&seed_digest)?;
-        seed_digest[0].try_into()
-    }
+        let mut felts: [Felt; 2] = seed_digest.as_elements()[0..2]
+            .try_into()
+            .expect("we should have sliced off 2 elements");
 
-    /// Creates a new [AccountId] without checking its validity.
-    ///
-    /// This function requires that the provided value is a valid [Felt] representation of an
-    /// [AccountId].
-    pub fn new_unchecked(value: Felt) -> Self {
-        Self(value)
-    }
+        felts[1] = shape_second_felt(felts[1], anchor.epoch());
```

Comment:
Small question: Why do we grind the first felt metadata but the second felt we shape it? Would it be too hard of a PoW?